pkg/planner/core/hint_utils.go

```diff
@@ -220,59 +312,185 @@ func getTableName(tblName model.CIStr, asName *model.CIStr) model.CIStr {
+				hintTbls = append(hintTbls, nil)
+				continue
+			}
+			guessQBOffsets[qbOffset] = struct{}{}
+		}
+		qbOffsets = append(qbOffsets, qbOffset)
+		hintTbls = append(hintTbls, ht)
+	}
+
+	// 2. Add QB name for each table name in the hint.
+
+	for i, hintTbl := range hintTbls {
+		if hintTbl == nil {
+			continue
+		}
+		// In quick binding, we always put the generated hints in the first valid place in the SQL.
+		// That implies hintname(@del_1) and hintname(@upd_1) is unnecessary in UPDATE/DELETE statements, and
+		// hintname(@sel_1) is unnecessary in SELECT statements.
+		// We don't generate QB name for the table names in the hint in this case to make the result cleaner.
+		if (qbOffsets[i] <= 1 && nodeType == h.TypeSelect) ||
+			(qbOffsets[i] == 0 && (nodeType == h.TypeUpdate || nodeType == h.TypeDelete)) {
 			continue
 		}
-		qbName, err := h.GenerateQBName(nodeType, qbOffset)
+		tblQBName, err := h.GenerateQBName(nodeType, qbOffsets[i])
 		if err != nil {
 			continue
 		}
-		return &ast.TableOptimizerHint{
-			QBName:   qbName,
-			HintName: model.NewCIStr(joinType),
-			Tables:   []ast.HintTable{*ht},
+		hintTbls[i].QBName = tblQBName
+	}
+
+	// 3. Generate QB name for the hint itself based on the QB name of each join node from step 1.
+
+	// Current join reorder will break QB offset of the join operator, e.g. setting them to -1.
+	// So we are unable to get the correct QB offset for the hint from the join operator, now we use the minimum QB
+	// offset among the tables.
+	// Besides, genHintTblForSingleJoinNode() is not powerful enough to handle all cases, it may fail in some cases.
+	// If we failed to get QB offset information from one join node, we don't generate QB name for the hint. Because
+	// that may cause a wrong QB offset, leaving it blank is probably better.
+	if slices.Contains(qbOffsets, -1) {
+		return hintTbls, nil
+	}
+	minQBOffset := slices.Min(qbOffsets)
+
+	// ditto. We don't generate unnecessary QB name for the hint itself.
+	if (minQBOffset > 1 && nodeType == h.TypeSelect) ||
+		(minQBOffset > 0 && (nodeType == h.TypeUpdate || nodeType == h.TypeDelete)) {
+		hintQBName, err := h.GenerateQBName(nodeType, minQBOffset)
+		if err != nil {
+			return nil, nil
 		}
+		hintQBNamePtr = &hintQBName
 	}
-	return res
+	return hintTbls, hintQBNamePtr
 }
 
-func extractHintTableForJoinNode(
+// genHintTblForSingleJoinNode tries to generate ast.HintTable and QB offset for a single join node.
+// See the comments inside about the meaning of guessQBOffset.
+func genHintTblForSingleJoinNode(
 	sctx base.PlanContext,
 	joinNode base.PhysicalPlan,
 	parentOffset int,
 ) (
 	qbOffset int,
+	guessQBOffset bool,
 	ht *ast.HintTable,
 ) {
-	qbOffset = joinNode.QueryBlockOffset()
+	selfOffset := joinNode.QueryBlockOffset()
+	qbOffset = selfOffset
 	if qbOffset == -1 {
-		return -1, nil
+		return -1, false, nil
 	}
+	guessQBOffset = false
 	var dbName, tableName *model.CIStr
 	if qbOffset != parentOffset {
```

Comment:
This can only happen when the joined node is from a sub-query? Better to add some comments.